File: blind75/02_best_time_to_buy_and_sell_stock/best_time_to_buy_and_sell_stock.py
from typing import List


def max_profit(prices: List[int]) -> int:
    """
    Returns the maximum profit from buying and selling stock once.
    Uses single pass with O(n) time and O(1) space.

    Args:
        prices: List of stock prices where prices[i] is the price on day i

    Returns:
        Maximum profit achievable, or 0 if no profit is possible
    """
    if not prices:
        return 0

    min_price = prices[0]
    max_profit = 0

    for price in prices:
        if price < min_price:
            min_price = price
        elif price - min_price > max_profit:
            max_profit = price - min_price

    return max_profit

File: blind75/02_best_time_to_buy_and_sell_stock/best_time_to_buy_and_sell_stock.go
package best_time_to_buy_and_sell_stock

// MaxProfit returns the maximum profit from buying and selling stock once.
// Uses single pass with O(n) time and O(1) space.
func MaxProfit(prices []int) int {
	if len(prices) == 0 {
		return 0
	}

	minPrice := prices[0]
	maxProfit := 0

	for _, price := range prices {
		if price < minPrice {
			minPrice = price
		} else if profit := price - minPrice; profit > maxProfit {
			maxProfit = profit
		}
	}

	return maxProfit
}
